File: back/controllers/groups/groupController.js
const Group = require('../../models/groups/groupModel');
const User = require('../../models/users/userModel');
const GroupMessage = require('../../models/groups/groupMessageModel');

// Create a new group
exports.createGroup = async (req, res) => {
  try {
    const { groupName, participants, description, groupImageUrl } = req.body;
    const userId = req.user.id;

    // Add the creator to the participants list if not already included
    const allParticipants = participants.includes(userId) ? participants : [...participants, userId];

    // Validate that all participants exist
    const validUsers = await User.find({ _id: { $in: allParticipants } });
    if (validUsers.length !== allParticipants.length) {
      return res.status(400).json({ error: 'Some participants do not exist' });
    }

    const group = new Group({
      participants: allParticipants,
      createdBy: userId,
      groupName,
      description,
      groupImageUrl,
      unreadCounts: new Map()
    });

    await group.save();
    await group.populate('participants', 'name email profileImageUrl');

    res.status(200).json({ success: true, group });
  } catch (error) {
    console.error('Error creating group:', error);
    res.status(500).json({ error: 'Failed to create group' });
  }
};

// Get all groups for the current user
exports.getUserGroups = async (req, res) => {
  try {
    const userId = req.user.id;

    const groups = await Group.find({
      participants: userId
    })
      .populate('participants', 'name email profileImageUrl')
      .populate('lastMessage')
      .sort({ updatedAt: -1 });

    res.status(200).json({ groups });
  } catch (error) {
    console.error('Error fetching groups:', error);
    res.status(500).json({ error: 'Failed to fetch groups' });
  }
};

// Get all groups (where user is not a participant) 
exports.getAllGroups = async (req, res) => {
  try {
    const groups = await Group.find()
      .populate('participants', 'name email profileImageUrl')
      .populate('lastMessage')
      .sort({ updatedAt: -1 });

    res.status(200).json({ groups });
  } catch (error) {
    console.error('Error fetching groups:', error);
    res.status(500).json({ error: 'Failed to fetch groups' });
  }
};

// Get group details
exports.getGroupDetails = async (req, res) => {
  try {
    const { groupId } = req.params;
    const userId = req.user.id;

    const group = await Group.findById(groupId)
      .populate('participants', 'name email profileImageUrl')
      .populate('lastMessage');

    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is a participant
    if (!group.participants.some(participant => participant._id.toString() === userId)) {
      return res.status(403).json({ error: 'Access denied' });
    }

    res.status(200).json({ group });
  } catch (error) {
    console.error('Error getting group details:', error);
    res.status(500).json({ error: 'Failed to get group details' });
  }
};

// Update group details
exports.updateGroup = async (req, res) => {
  try {
    const { groupId } = req.params;
    const { groupName, description, groupImageUrl } = req.body;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Update fields if provided
    group.groupName = groupName;
    group.description = description;
    group.groupImageUrl = groupImageUrl;
    console.log('Updated group:', group);
    await group.save();
    await group.populate('participants', 'name email profileImageUrl');

    res.status(200).json({ success: true, group });
  } catch (error) {
    console.error('Error updating group:', error);
    res.status(500).json({ error: 'Failed to update group' });
  }
};

// Delete a group
exports.deleteGroup = async (req, res) => {
  try {
    const { groupId } = req.params;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is the creator or admin
    if (group.createdBy.toString() !== userId) {
      return res.status(403).json({ error: 'Only group creator can delete the group' });
    }

    // Delete all group messages
    await GroupMessage.deleteMany({ groupId });

    // Delete the group
    await Group.findByIdAndDelete(groupId);

    res.status(200).json({ success: true, message: 'Group deleted successfully' });
  } catch (error) {
    console.error('Error deleting group:', error);
    res.status(500).json({ error: 'Failed to delete group' });
  }
};

// Leave a group
exports.leaveGroup = async (req, res) => {
  try {
    const { groupId } = req.params;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is a participant
    if (!group.participants.includes(userId)) {
      return res.status(404).json({ error: 'User is not a member of this group' });
    }

    // Remove user from participants
    group.participants = group.participants.filter(
      participantId => participantId.toString() !== userId
    );

    // Remove user from unread counts
    group.unreadCounts.delete(userId);

    await group.save();

    res.status(200).json({ success: true, message: 'Left group successfully' });
  } catch (error) {
    console.error('Error leaving group:', error);
    res.status(500).json({ error: 'Failed to leave group' });
  }
};

// Add participants to group
exports.addParticipants = async (req, res) => {
  try {
    const { groupId } = req.params;
    const { participants } = req.body;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Validate that all new participants exist
    const validUsers = await User.find({ _id: { $in: participants } });
    if (validUsers.length !== participants.length) {
      return res.status(400).json({ error: 'Some participants do not exist' });
    }

    // Add new participants (avoid duplicates)
    const newParticipants = participants.filter(
      participantId => !group.participants.includes(participantId)
    );

    group.participants.push(...newParticipants);
    await group.save();
    await group.populate('participants', 'name email profileImageUrl');
    res.status(200).json({ success: true, group });
  } catch (error) {
    console.error('Error adding participants:', error);
    res.status(500).json({ error: 'Failed to add participants' });
  }
};

// Remove participant from group
exports.removeParticipant = async (req, res) => {
  try {
    const { groupId } = req.params;
    const { participantId } = req.body;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is the creator or admin
    if (group.createdBy.toString() !== userId) {
      return res.status(403).json({ error: 'Only group creator can remove participants' });
    }

    // Check if participant exists in group
    if (!group.participants.includes(participantId)) {
      return res.status(404).json({ error: 'Participant not found in group' });
    }

    // Remove participant
    group.participants = group.participants.filter(
      id => id.toString() !== participantId
    );

    // Remove from unread counts
    group.unreadCounts.delete(participantId);

    await group.save();
    await group.populate('participants', 'name email profileImageUrl');

    res.status(200).json({ success: true, group });
  } catch (error) {
    console.error('Error removing participant:', error);
    res.status(500).json({ error: 'Failed to remove participant' });
  }
};

// Make participant admin (you'll need to add admin field to group model)
exports.makeAdmin = async (req, res) => {
  try {
    const { groupId } = req.params;
    const { participantId } = req.body;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is the creator
    if (group.createdBy.toString() !== userId) {
      return res.status(403).json({ error: 'Only group creator can make admins' });
    }

    // Check if participant exists in group
    if (!group.participants.includes(participantId)) {
      return res.status(404).json({ error: 'Participant not found in group' });
    }

    // Add admin functionality here when you add admins field to group model
    // For now, just return success
    res.status(200).json({ success: true, message: 'Admin functionality to be implemented' });
  } catch (error) {
    console.error('Error making admin:', error);
    res.status(500).json({ error: 'Failed to make admin' });
  }
};

// Remove admin privileges
exports.removeAdmin = async (req, res) => {
  try {
    const { groupId } = req.params;
    const { participantId } = req.body;
    const userId = req.user.id;

    const group = await Group.findById(groupId);
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is the creator
    if (group.createdBy.toString() !== userId) {
      return res.status(403).json({ error: 'Only group creator can remove admin privileges' });
    }

    // Add admin functionality here when you add admins field to group model
    // For now, just return success
    res.status(200).json({ success: true, message: 'Remove admin functionality to be implemented' });
  } catch (error) {
    console.error('Error removing admin:', error);
    res.status(500).json({ error: 'Failed to remove admin' });
  }
};

// Search groups
exports.searchGroups = async (req, res) => {
  try {
    const { q } = req.query;
    const userId = req.user.id;

    if (!q || q.trim() === '') {
      return res.status(400).json({ error: 'Search query is required' });
    }

    const groups = await Group.find({
      participants: userId,
      $or: [
        { groupName: { $regex: q, $options: 'i' } },
        { description: { $regex: q, $options: 'i' } }
      ]
    })
      .populate('participants', 'name email profileImageUrl')
      .populate('lastMessage')
      .sort({ updatedAt: -1 });

    res.status(200).json({ groups });
  } catch (error) {
    console.error('Error searching groups:', error);
    res.status(500).json({ error: 'Failed to search groups' });
  }
};

// Get group participants
exports.getGroupParticipants = async (req, res) => {
  try {
    const { groupId } = req.params;
    const userId = req.user.id;

    const group = await Group.findById(groupId)
      .populate('participants', 'name email profileImageUrl');

    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Check if user is a participant
    if (!group.participants.some(participant => participant._id.toString() === userId)) {
      return res.status(403).json({ error: 'Access denied' });
    }

    res.status(200).json({ participants: group.participants });
  } catch (error) {
    console.error('Error getting group participants:', error);
    res.status(500).json({ error: 'Failed to get group participants' });
  }
};

File: front/lib/controller/groups/group_controller.dart
import 'dart:convert';
import 'package:front/main.dart';
import 'package:front/model/groups/group_model.dart';
import 'package:front/model/users/user_model.dart';
import 'package:front/services/socket_service.dart';
import 'package:front/utils/hive_utils.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

class GroupController {
  // ==================== GROUP MANAGEMENT ====================
  // Note: Message-related functionalities (including polls) are in GroupMessageController

  /// Fetch all groups for the current user
  static Future<List<Group>> fetchUserGroups() async {
    try {
      final token = HiveUtils.getData('token');
      print(token);
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.get(
        Uri.parse('$apiUrl/group/getGroups'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final dynamic decodedResponse = json.decode(response.body);

        // Handle both possible formats: direct list or map with 'groups' key
        List<dynamic> data;
        if (decodedResponse is List) {
          data = decodedResponse;
        } else if (decodedResponse is Map<String, dynamic>) {
          data = decodedResponse['groups'] ?? [];
        } else {
          throw Exception('Unexpected response format');
        }
        debugPrint('Fetched groups: $data');
        if (data.isNotEmpty) {
          debugPrint('Fetched groups count: ${data.length}, first group: ${data[0]}');
        } else {
          debugPrint('No groups found');
        }

        return data.map((group) => Group.fromJson(group)).toList();
      } else {
        throw Exception('Failed to load groups: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error fetching groups: $e');
      throw Exception('Failed to load groups: $e');
    }
  }

  /// Fetch all groups
  static Future<List<Group>> fetchGroups() async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.get(
        Uri.parse('$apiUrl/group/getAllGroups'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        debugPrint('Response body: ${response.body}');
        final List<dynamic> jsonData = json.decode(response.body)['groups'];
        return jsonData.map((group) => Group.fromJson(group)).toList();
      } else {
        throw Exception('Failed to load groups: ${response.statusCode}');
      }
    } catch (e, stackTrace) {
      debugPrint('Error fetching groups: $e');
      debugPrint('Stack trace: $stackTrace');
      throw Exception('Failed to load groups: $e');
    }
  }

  /// Create a new group
  static Future<bool> createGroup({
    required String groupName,
    required List<String> participantIds,
    String? description,
    String? groupImageUrl,
  }) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$apiUrl/group/createGroup'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({
          'groupName': groupName,
          'participants': participantIds,
          'description': description,
          'groupImageUrl': groupImageUrl,
        }),
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to create group: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error creating group: $e');
      throw Exception('Failed to create group: $e');
    }
  }

  /// Update group details
  static Future<bool> updateGroup({
    required String groupId,
    String? groupName,
    String? description,
    String? groupImageUrl,
  }) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      debugPrint('Updating group with image URL: $groupImageUrl');

      final response = await http.patch(
        Uri.parse('$apiUrl/group/$groupId'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({
          'groupName': groupName,
          'description': description,
          'groupImageUrl': groupImageUrl,
        }),
      );

      debugPrint('Update group response: ${response.body}');

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to update group: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error updating group: $e');
      throw Exception('Failed to update group: $e');
    }
  }

  /// Delete a group
  static Future<bool> deleteGroup(String groupId) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.delete(
        Uri.parse('$apiUrl/group/$groupId'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to delete group: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error deleting group: $e');
      throw Exception('Failed to delete group: $e');
    }
  }

  /// Leave a group
  static Future<bool> leaveGroup(String groupId) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$apiUrl/group/$groupId/leave'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to leave group: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error leaving group: $e');
      throw Exception('Failed to leave group: $e');
    }
  }

  // ==================== PARTICIPANT MANAGEMENT ====================

  /// Add participants to group
  static Future<bool> addParticipants({
    required String groupId,
    required List<String> participantIds,
  }) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$apiUrl/group/$groupId/addParticipants'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({'participants': participantIds}),
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to add participants: ${response.statusCode}');
      }
    } catch (e, stackTrace) {
      debugPrint('Error adding participants: $e');
      debugPrint('Stack trace: $stackTrace');
      throw Exception('Failed to add participants: $e');
    }
  }

  /// Remove participant from group
  static Future<bool> removeParticipant({
    required String groupId,
    required String participantId,
  }) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$apiUrl/group/$groupId/removeParticipant'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({'participantId': participantId}),
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to remove participant: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error removing participant: $e');
      throw Exception('Failed to remove participant: $e');
    }
  }

  /// Make participant admin
  static Future<bool> makeAdmin({
    required String groupId,
    required String participantId,
  }) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$apiUrl/group/$groupId/makeAdmin'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({'participantId': participantId}),
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to make admin: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error making admin: $e');
      throw Exception('Failed to make admin: $e');
    }
  }

  /// Remove admin privileges
  static Future<bool> removeAdmin({
    required String groupId,
    required String participantId,
  }) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$apiUrl/group/$groupId/removeAdmin'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({'participantId': participantId}),
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to remove admin: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error removing admin: $e');
      throw Exception('Failed to remove admin: $e');
    }
  }

  // ==================== SEARCH AND UTILITY ====================

  /// Search groups
  static Future<List<Group>> searchGroups(String query) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.get(
        Uri.parse('$apiUrl/group/search?q=${Uri.encodeComponent(query)}'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final dynamic decodedResponse = json.decode(response.body);

        List<dynamic> data;
        if (decodedResponse is List) {
          data = decodedResponse;
        } else if (decodedResponse is Map<String, dynamic>) {
          data = decodedResponse['groups'] ?? [];
        } else {
          throw Exception('Unexpected response format');
        }

        return data.map((group) => Group.fromJson(group)).toList();
      } else {
        throw Exception('Failed to search groups: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error searching groups: $e');
      throw Exception('Failed to search groups: $e');
    }
  }

  /// Get group participants
  static Future<List<User>> getGroupParticipants(String groupId) async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.get(
        Uri.parse('$apiUrl/group/$groupId/participants'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final dynamic decodedResponse = json.decode(response.body);

        List<dynamic> data;
        if (decodedResponse is List) {
          data = decodedResponse;
        } else if (decodedResponse is Map<String, dynamic>) {
          data = decodedResponse['participants'] ?? [];
        } else {
          throw Exception('Unexpected response format');
        }

        return data.map((user) => User.fromJson(user)).toList();
      } else {
        throw Exception(
          'Failed to get group participants: ${response.statusCode}',
        );
      }
    } catch (e) {
      debugPrint('Error getting group participants: $e');
      throw Exception('Failed to get group participants: $e');
    }
  }

  // ==================== SOCKET HELPERS ====================

  /// Join group room for real-time updates
  static void joinGroupRoom(String groupId) {
    final socketService = SocketService();
    if (socketService.isConnected) {
      socketService.socket.emit('join_group', {'groupId': groupId});
    }
  }

  /// Leave group room
  static void leaveGroupRoom(String groupId) {
    final socketService = SocketService();
    if (socketService.isConnected) {
      socketService.socket.emit('leave_group', {'groupId': groupId});
    }
  }

  /// Get available users for group creation
  static Future<List<User>> getAvailableUsers() async {
    try {
      final token = HiveUtils.getData('token');
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.get(
        Uri.parse(
          '$apiUrl/users/all',
        ), // Assuming there's an endpoint to get all users
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final dynamic decodedResponse = json.decode(response.body);

        List<dynamic> data;
        if (decodedResponse is List) {
          data = decodedResponse;
        } else if (decodedResponse is Map<String, dynamic>) {
          data = decodedResponse['users'] ?? [];
        } else {
          throw Exception('Unexpected response format');
        }

        return data.map((user) => User.fromJson(user)).toList();
      } else {
        throw Exception('Failed to get users: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('Error getting users: $e');
      throw Exception('Failed to get users: $e');
    }
  }
}
